### ISSUE TITLE
Calibration between actual retention and predicted retention is not great

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs4anki#215]
Btw, calibration doesn't look great on my collection. I've tried it with separate decks and got similar results
![Calibration graph (Entire collection)](https://user-images.githubusercontent.com/83031600/231520566-8039c37c-599d-458d-b933-5c8f5907a4ca.png)

_Originally posted by @Expertium in https://github.com/open-spaced-repetition/fsrs4anki/issues/151#issuecomment-1505591507_

**Comment by user1823** [comment]
This is also confirmed by the following stats:

<img width="395" alt="image" src="https://user-images.githubusercontent.com/92206575/232238586-987f259e-306d-400a-ba56-925ee4987ebb.png">

**Comment by Expertium** [comment]
Wait, I don't get it. How is it confirmed by these stats?

**Comment by user1823** [comment]
FSRS predicts that the average retention for the cards should be 96.85%.

But, over the past one month, the retention was just 94.2%

**Comment by Expertium** [comment]
Nah, that's a pretty small discrepancy. What FSRS predicts and what True Retention shows you are somewhat different things (I wouldn't ask Sherlock to implement it if I could just use True Retention instead, they're not identical), so a small discrepancy like this is fine. If it was like 70% vs 95%, that would be worrying.

**Comment by L-M-Sherlock** [comment]
> FSRS predicts that the average retention for the cards should be 96.85%.
> 
> But, over the past one month, the measured retention was just 94.2%

They are not the same thing. The retention in FSRS stats is calculated in all cards including those undue cards. The retention of undue cards is higher than your requested retention.

**Comment by user1823** [comment]
> They are not the same thing. The retention in FSRS stats is calculated in all cards including those undue cards. The retention of undue cards is higher than your requested retention.

Oh! I failed to consider that.

So, should I close this issue?

**Comment by user1823** [comment]
> The retention in FSRS stats is calculated in all cards including those undue cards. The retention of undue cards is higher than your requested retention.

But, I think that you should include this in the explanation for the average retention stats when you release the feature.

**Comment by L-M-Sherlock** [comment]
OK, I will improve the explanation of the FSRS stats.

**Comment by user1823** [comment]
> In my recent research (unpublished), the error of power functions is lower than that of exponential functions in fitting the forgetting curve. I will test it in some users' collections.

_Originally posted by @L-M-Sherlock in https://github.com/open-spaced-repetition/fsrs4anki/issues/151#issuecomment-1506399330_

> Interesting. Woz says that the forgetting curve is definitely exponential, but when you have both difficult and easy material in the same collection, the resulting curve (a superposition of different exponential curves) is best approximated with a power function.
> https://supermemo.guru/wiki/Exponential_nature_of_forgetting#Power_law_emerges_in_superposition_of_exponential_forgetting_curves

_Originally posted by @Expertium in https://github.com/open-spaced-repetition/fsrs4anki/issues/151#issuecomment-1506415676_

@L-M-Sherlock, you may want to test the power function in my collection. I believe that my collection includes significant amounts of both easy and difficult materials.

- My deck: [Default.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/11311507/Default.zip) (Change extension to .apkg)
- w: [1.0696, 1.6707, 4.9898, -1.2855, -1.1631, 0.0, 1.7095, -0.0887, 1.0651, 1.71, -0.4937, 0.8296, 0.4661]
- requestRetention: 0.94
- maximumInterval: 36500
- easyBonus: 2.5
- hardInterval: 1.2
- timezone = 'Asia/Calcutta'
- next_day_starts_at = 4
- revlog_start_date = "2006-10-05"

**Comment by L-M-Sherlock** [comment]
Thanks for the data. I will do some research here. Currently I am working on filtering out the outliers in the data of Expertium.

**Comment by user1823** [comment]
@L-M-Sherlock, please let me know the results after you test the power function in my collection. I think that it might solve [this issue](https://github.com/open-spaced-repetition/fsrs4anki/issues/187) also (at least partially).

**Comment by L-M-Sherlock** [comment]
Unfortunately, I even increased the number of parameters to 20 today, but the accuracy doesn't increase significantly.

**Comment by L-M-Sherlock** [comment]
Unfortunately, I even increased the number of parameters to 20 today, but the accuracy doesn't increase significantly. I need to do more experiments here.

**Comment by Expertium** [comment]
I'm curious what parameters you added. Do you mind giving a detailed description of this new model, with all the formulas?

**Comment by L-M-Sherlock** [comment]
I will share some details about it tomorrow.

**Comment by L-M-Sherlock** [comment]
Add power forgetting curve and power difficulty:

![image](https://user-images.githubusercontent.com/32575846/234485630-69c6568e-3650-405f-88be-be13f396df6e.png)

I think there is not significant difference.

![image](https://user-images.githubusercontent.com/32575846/234485686-4cba5878-4505-4861-8901-5dd3db3091e9.png)

**Comment by L-M-Sherlock** [comment]
> Previously you said that you have increased the number of parameters to 20. Perhaps you could release a beta-version of the optimizer and a beta-version of the scheduler with those parameters so other people can experiment with them (and do side-by-side comparisons) as well?

OK, I will publish the beta-version of the optimizer in another branch. But I don't want to develop the scheduler until the optimizer is significant better than before and stable.

**Comment by Expertium** [comment]
Also, a suggestion - add root-mean-square error too, not just R^2.
![download](https://user-images.githubusercontent.com/83031600/234519947-14f8f89c-d991-4a81-8977-aa6545dbe3a4.png)
1. Find the difference between "Average actual retention" and "Optimal average actual retention", and square it
2. Take the sum of all squared differences, divide it by their count, and take a square root.
3. The interpretation is as follows: it tells us how much, on average, the calibration is off from theoretical perfection. For example, RMSE=0.05 means that on average FSRS is 5% off from theoretically perfect calibration. This is easier to interpret than R^2, though I think we should keep both.

**Comment by Expertium** [comment]
Also, maybe I'm just dumb, but I have no idea what is going on here.
![image](https://user-images.githubusercontent.com/83031600/234560317-f9e00764-a331-4dc9-b457-27240a23c067.png)
Is this correct? If so, why is there a 9, and no new parameters?
![image](https://user-images.githubusercontent.com/83031600/234560528-464c477d-4868-48f2-bd83-9470d98a38a5.png)

**Comment by user1823** [comment]
> I thought the power will be a new parameter, like this.
> ![image](https://user-images.githubusercontent.com/83031600/234562199-d6e32afd-e455-4071-914d-24079b1cd9e3.png)

In the following image (from SuperMemo), the power function is of the form:
Retrievability = $A t^B$

<img src="https://user-images.githubusercontent.com/92206575/234565713-f7fb4c99-f6ca-415f-8300-109f47df7123.png">

So, I think that the power in the equation shared by @L-M-Sherlock above is not what we are targeting here.

**Comment by L-M-Sherlock** [comment]
![image](https://user-images.githubusercontent.com/32575846/234577035-8f0ccda1-30f1-4cb8-a8db-b331d89cfccc.png)

I have to ensure that stability has the same meaning as before. You can compare the f(t) and g(t) in the graph. when t = a, f(a) = g(a) = 0.9.

**Comment by user1823** [comment]
I am not sure that I understand the algorithm correctly. But,

If the original function is
$$R = A~e^{-kt}$$

shouldn't the formula for the retention in terms of stability be
$$R = 0.9~e^{-k(t-S)}$$

How did it become the following?
$$R = 0.9^{t/S}$$


And now, if the new function is
$$R = A~t^B$$

shouldn't the formula for the retention in terms of stability be
$$R = 0.9~(t/S)^B$$

How did it become the following?
$$R = (1 + \frac{t}{9S})^{-1}$$

**Comment by L-M-Sherlock** [comment]
There is only one parameter for the forgetting curve function. The exponential forgetting curve function is $R(t)=e^{\ln(0.9)t/S}$, which is equivalent to $R(t)=0.9^{t/S}$.

**Comment by user1823** [comment]
So, is this some other formula?

The two formulas I mention are from the image posted above. The source of the image is https://supermemo.guru/wiki/Exponential_nature_of_forgetting#Power_law_emerges_in_superposition_of_exponential_forgetting_curves.

**Comment by L-M-Sherlock** [comment]
In Algorithm SM-17, retrievability R corresponds with the probability of recall and represents the exponential forgetting curve. Retrievability is derived from stability and the interval:

R[n]:=exp-k*t/S[n-1]

where:

R[n] - retrievability at the n-th repetition
k - decay constant
t - time (interval)
S[n-1] - stability after the (n-1)th repetition

**Comment by user1823** [comment]
The new formula is similar to the one below. 
> Wixted and Carpenter (2007) summarize the forgetting curve as:
P(recall) = $m(1 + ht)^{-f}$
Where m is the degree of initial learning (i.e. the probability at time 0), h is a scaling factor on time, and f is the exponential memory decay factor.

Source: https://notes.andymatuschak.org/zHdKY3GwoUW9xG6wQtKFqjz9jcrxdM3mxram

So, I think that like what @Expertium says, if the power is -w instead of -1, the results would be more impressive.

**Comment by L-M-Sherlock** [comment]
OK, I will test it tomorrow.

**Comment by user1823** [comment]
@Expertium, you can try out the current version of the new optimizer here: https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Feat%2Fpow_difficulty/fsrs4anki_optimizer.ipynb

**Comment by Expertium** [comment]
I tried and got exactly the same results. Like, down to the 4th significant digit. Either I somehow opened the same version of the optimizer (and I'm sure I opened different version), or...I don't even know.
![New](https://user-images.githubusercontent.com/83031600/234631298-2924a038-0329-4eab-8985-8461f0f86180.png)
![Old](https://user-images.githubusercontent.com/83031600/234631307-2274e070-022d-49de-87bf-2ed508537025.png)

**Comment by user1823** [comment]
Interesting. I got very different results from the two. But, the result from the original optimizer was better.

Old optimizer:
![calibration original optimizer](https://user-images.githubusercontent.com/92206575/234634158-533f642b-5ba4-4536-a6db-d97f358d187e.png)
![log loss original optimizer](https://user-images.githubusercontent.com/92206575/234634192-629ddb8f-c95b-47ec-b428-2e342a356d44.png)

New optimizer:
![calibration new optimizer 26-04](https://user-images.githubusercontent.com/92206575/234634338-50041b09-3a0c-4bdd-bf35-dd688b83db21.png)
![log loss new optimizer 26-04](https://user-images.githubusercontent.com/92206575/234634372-e1e060cb-4031-494c-a44b-8fd4bf8f0bd5.png)

**Comment by kieranlblack** [comment]
It doesn't look like all of the changes LM was displaying earlier are up in the github, I don't have time at the present to look at things more closely, but this is likely why you are seeing no change.

**Comment by L-M-Sherlock** [comment]
> Interesting. I got very different results from the two. But, the result from the original optimizer was better.

It even has more parameters:
```python
init_s = [1, 2, 3, 4]
init_d = [4, 3, 2, 1]
init_o = [2, 1, 0, -1]
init_w = [0.2, 1.4, -0.12, 0.8, 2, -0.2, 0.2, 1]
```
But it become worse than before.

**Comment by L-M-Sherlock** [comment]
> The new formula is similar to the one below.
> 
> > Wixted and Carpenter (2007) summarize the forgetting curve as:
> > P(recall) = m(1+ht)−f
> > Where m is the degree of initial learning (i.e. the probability at time 0), h is a scaling factor on time, and f is the exponential memory decay factor.
> 
> Source: https://notes.andymatuschak.org/zHdKY3GwoUW9xG6wQtKFqjz9jcrxdM3mxram
> 
> So, I think that like what @Expertium says, if the power is -w instead of -1, the results would be more impressive.

It work well for me:

<img width="1235" alt="image" src="https://user-images.githubusercontent.com/32575846/234795012-31ae4658-2e3b-49b7-984d-e6b777b0c635.png">

Commit: https://github.com/open-spaced-repetition/fsrs4anki/commit/301de8f57f3402210edec89aa3dd2036e2061fac

You can test it for yourself: https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Expt/power-forgetting-curve/fsrs4anki_optimizer.ipynb

**Comment by Expertium** [comment]
Also, I still don't understand the meaning of 9 here:
![image](https://user-images.githubusercontent.com/83031600/234821214-bc87b960-1824-480c-8d62-1f42606bb19a.png)
@L-M-Sherlock, can you explain where 9 comes from in this formula?

**Comment by L-M-Sherlock** [comment]
> This shouldn't be possible by the very definiton of R^2 - it's a square of correlation coefficient, a square cannot be negative.

It can be negative. I use the standard implementation from sklearn.

**Comment by L-M-Sherlock** [comment]
> , can you explain where 9 comes from in this formula?

It's just a constant to ensure that the new stability has similar scale as before.

**Comment by L-M-Sherlock** [comment]
I think R-squared is not a meaningful metric in this case. I don't know how woz calculate it in his graph: https://supermemo.guru/wiki/Universal_metric_for_cross-comparison_of_spaced_repetition_algorithms#Algorithm_SM-17_as_a_good_classifier

**Comment by Expertium** [comment]
It is meaningful (though RMSE is better IMO), and your previous implementation worked fine, it just seems that the new implementation has a bug.

**Comment by L-M-Sherlock** [comment]
> There are cases where R2 can yield negative values. This can arise when the predictions that are being compared to the corresponding outcomes have not been derived from a model-fitting procedure using those data. Even if a model-fitting procedure has been used, R2 may still be negative, for example when linear regression is conducted without including an intercept,[[5]](https://en.wikipedia.org/wiki/Coefficient_of_determination#cite_note-5) or when a non-linear function is used to fit the data.[[6]](https://en.wikipedia.org/wiki/Coefficient_of_determination#cite_note-6) In cases where negative values arise, the mean of the data provides a better fit to the outcomes than do the fitted function values, according to this particular criterion.

Source: https://en.wikipedia.org/wiki/Coefficient_of_determination

**Comment by Expertium** [comment]
Huh. Interesting, I didn't know that.

Alright, I just won't use R^2 and focus on log-loss and RMSE when comparing decks.

**Comment by Expertium** [comment]
I tested it on 7 different decks with varying subjective difficulty of the material, plus the entire collection, so 8 data points in total.
For each deck both loss and RMSE were lower when using the new version of the optimizer. On average, log-loss was around 4% lower, and RMSE was around 16% lower.
What's interesting is that when I ran the new one on the full collection, RMSE was **lower** than when I ran it on individual decks, which suggests that maybe having different parameters for different decks is not actually advantageous.
![image](https://user-images.githubusercontent.com/83031600/234875228-704ac583-814b-4dca-a7ad-961f19c2820c.png)
It's also interesting how the shape of the blue curve changed entirely.

Next I ran a fancy test of statistical significane (called Wilcoxon signed-rank test) on both loss data and RMSE data to check that it's not just a coincidence, and got a p-value <0.01 in both cases, which means that we can be sure that both loss and RMSE are lower with the new optimizer and it's not a fluke.

Also, (while it's a bit late to ask) is pre-training in 2.2 really necessary? It seems like a very non-intuitive thing to me.

I suppose the next step is to apply the same idea to difficulty and replace (10 * torch.pow(new_d, -1)) with something like (10 * torch.pow(new_d, -f_2)). Also, for R^2 I think you should bring back the implementation you used in some of the previous versions, there is no way the values given by sklearn.metrics.r2_score are correct, somethign weird is going on with it.

**Comment by L-M-Sherlock** [comment]
> Also, (while it's a bit late to ask) is pre-training in 2.2 really necessary? It seems like a very non-intuitive thing to me.

Because I freeze the the parameters of initial stability in the training stage. It make the parameters more stable and accurate during the training.

**Comment by L-M-Sherlock** [comment]
> Also, for R^2 I think you should bring back the implementation you used in some of the previous versions, there is no way the values given by sklearn.metrics.r2_score are correct, something weird is going on with it.

The previous implementation is incorrect. It is not R-squared, because R-squared is not r^2. I believe the sklearn (the most popular machine learning package) implementation is correct.

**Comment by Expertium** [comment]
> because R-squared is not r^2

https://en.wikipedia.org/wiki/Coefficient_of_determination

![image](https://user-images.githubusercontent.com/83031600/234883540-f55d4d41-82ca-4d5e-bd67-5977e3a1f70e.png)

**Comment by galantra** [comment]
> What's interesting is that when I ran the new optimizer on the full collection, RMSE was **lower** than when I ran it on individual decks, which suggests that maybe having different parameters for different decks is not actually advantageous.

I noticed the same some time ago, IIRC. I imagine it is because of the data set is bigger, and a bigger data set will have a bigger impact vs. the initial Maimemo(?) data than a smaller data set would have. The Optimizer becomes more confident, so to say.

Is this correct?

**Comment by L-M-Sherlock** [comment]
> I suppose the next step is to apply the same idea to difficulty and replace (10 * torch.pow(new_d, -1)) with something like (10 * torch.pow(new_d, -f_2)) where f_2 is another parameter.

Nice idea! I will add it tomorrow.

**Comment by Expertium** [comment]
Despite the fact that the specific optimizer you are using is Adam (which has a lot of heuristics to make it more adaptive), learning rate still affects the results. I changed learning rate to lr = 5e-3 and got even better results with the new forgetting curve.
lr = 5e-4
Loss: 0.4589
RMSE: 0.0466

lr = 5e-3
Loss: 0.4478
RMSE: 0.0367

![image](https://user-images.githubusercontent.com/83031600/234914147-87760877-5c43-4682-a3ac-4845a8bca15a.png)

**Comment by user1823** [comment]
Also, increasing the n_epoch from 1 to 3 produces better results.

n_epoch = 1
<img width=50% src="https://user-images.githubusercontent.com/92206575/234918019-72caff89-da23-4c12-9762-73d3eb4b6242.png">
<img width=50% src="https://user-images.githubusercontent.com/92206575/234918029-6832981a-5185-4ca7-bca7-2c71ba3075f4.png">

n_epoch = 3
<img width=50% src="https://user-images.githubusercontent.com/92206575/234918027-1a9e667a-f13e-4e3d-9d2e-db30053ce531.png">
<img width=50% src="https://user-images.githubusercontent.com/92206575/234918034-143cf65a-40fd-40ab-b529-6a129837ffee.png">

**Comment by Expertium** [comment]
I tried changing learning rate to 5e-2 and got this error:
![image](https://user-images.githubusercontent.com/83031600/234924570-a286778a-96ae-4040-aeac-8fec9e65959a.png)

**Comment by L-M-Sherlock** [comment]
I'm stuck in a problem. The stability has different meaning for various users when I introduce `f` into the power forgetting curve:

<img width="1018" alt="image" src="https://user-images.githubusercontent.com/32575846/235085413-130b87e3-b78b-4c6f-84c6-d816c0ad0511.png">

**Comment by Expertium** [comment]
I guess that's inevitable if we're introducing parameters to the forgetting curve. If we introduce new parameters to the calculation of S while keeping the original forgetting curve (which only depends on S and t, no parameters) this problem can be circumvented. But then we will most likely end up with a higher loss and RMSE due to the forgetting curve not being very flexible, unless we improve calculation of S so much that it will compensate for that. Basically, we could decrease the loss/RMSE either by changing how S is calculated or by adding parameters to _R=f(S)_. Or both.

I guess it depends on what philosophy do you want to adopt for FSRS
1. "Only decreasing the loss matters, regardless of how ad hoc and non-rigorous our model is, and interpretability can go to hell"
2. "Every value and every parameter must have a precise meaning, sacrificing rigor and interpretability of the model is unacceptable"

**Comment by L-M-Sherlock** [comment]
If we finally decide to use power forgetting curve, I prefer a fixed `f` which is an average with training results collected from many users.

**Comment by Expertium** [comment]
That's a good idea, though we will need enormous amounts of data, not just from 3-5 users.
Ok, remove the power forgetting curve for now, bring back the exponential curve, and change difficulty formula to (10 * torch.pow(new_d, -f)), where f can be optimized.

**Comment by L-M-Sherlock** [comment]
> I suggest trying out the new approach in a new branch. Leave the current branch (with the power forgetting curve and optimized power) as it is.

Here is the notebook with exponential curve and 10 * torch.pow(new_d, -f): https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Expt/power-function-for-difficulty/fsrs4anki_optimizer.ipynb

The branch: https://github.com/open-spaced-repetition/fsrs4anki/tree/Expt/power-function-for-difficulty

**Comment by user1823** [comment]
> The stability has different meaning for various users when I introduce f into the power forgetting curve.

By the way, this problem can be solved by replacing the constant 9 in
$$R = (1 + \frac{t}{9S})^{f}$$
by
$$\frac{10^{1/f}}{9^{1/f} - 10^{1/f}}$$

However, I am not sure how practical it is.

**Comment by Expertium** [comment]
> Edit: A further simplified version can be R=(1+t(0.91/f−1)S)f

Then we lose flexibility, in other words, changing f will have barely any effect on the shape of the forgetting curve.

**Comment by Expertium** [comment]
> Here is the notebook with exponential curve and 10 * torch.pow(new_d, -f): https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Expt/power-function-for-difficulty/fsrs4anki_optimizer.ipynb

I've only tried it on 3 decks so far, but the results aren't looking good. Loss is around 0.5% higher than before, RMSE is around 2% higher. If anything, this new version is worse.

**Comment by user1823** [comment]
According to my testing, the performance of this version is in between that of the original optimizer and the power forgetting curve one. The results are summarized below:

  | Log loss | RMSE | R-squared
-- | -- | -- | --
Original Optimizer | 0.2241 | 0.0191 | 0.8728
Power forgetting curve | 0.2231 | 0.0147 | 0.9234
Power difficulty | 0.2234 | 0.0170 | 0.9026

**Comment by Expertium** [comment]
I tested it on 7 decks + entire collection, same as before. On average, log-loss is 0.6% worse than with the old (no power functions, no new parameters) algorithm, RMSE is 2% worse. I ran Wilcoxon signed-rank test on this data as well, but got high (>0.05) p-values, indicating that there is no statistically significant difference between how the old algorithm and the new one perform, aka they are roughly the same.

**Comment by Expertium** [comment]
Also, unlike with the power forgetting curve, this time changing the learning rate didn't help.

**Comment by user1823** [comment]
Since the power difficulty approach didn't produce the desired results, we might need to either
- use the power forgetting curve with a fixed value of `f`; or
- continue using the exponential forgetting curve and think of other ways to improve the calculation of the stability.

> If we finally decide to use power forgetting curve, I prefer a fixed f which is an average with training results collected from many users.

To examine the feasibility of this approach, I guess that the first step would be to compare the optimized `f` values for some users' collections.

The optimized `f` value for my collection is `-0.5501`.

**Comment by Expertium** [comment]
Mine is -0.1671, but this isn't the proper way of doing this. We need to collect the number of reviews as well (not just the value of `f`), so we can take a weighted average. But mroe importantly, we need at least a few dozens of people to submit their data, ideally more than a hundred.

**Comment by Expertium** [comment]
I just thought about something. If we use R=(1+t/9*S)^-1 as a formula, then R=0.9 when S=1, so S has a meaning that is easy to explain: it's how many days it takes for your retention to drop from 100% to 90% (btw, I finally understood the meaning of 9 in that formula). However, if we change the power to any value other than 1, the meaning of S changes regardless of what this value is and regardless of whether it stays constant or no.
For example, if f=0.5, then the meaning of S is "the number of days it takes your retention to drop from 100% to 94.87%".
What I'm trying to say is that **it doesn't matter whether `f` is constant or can be changed**. So the "let's take values of `f` from different users and average them" approach will also mess up the meaning of S. It's slightly better than making f optimizable for each user since at least the meaning of S will be the same across all users if it's just an average, but it still deviates from the original meaning that @L-M-Sherlock intended from the beginning.
TLDR: if the formula is R=(1+t/9*S)^-f, then **`f` cannot be anything other than 1**, otherwise the meaning of S - number days it takes for user's retention to drop from 100% to 90% - will change, regardless of whether it's a constant or an optimizable parameter.

**Comment by L-M-Sherlock** [comment]
Upon my recent thinking, I have resolved not to adopt any ideas that involve adding parameters on the forgetting curve. My reasoning is as follows: 

The optimal method for fitting the forgetting curve is given the same review history, reviewing at different intervals, calculating the retention rate, and subsequently plotting the following graph:

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/1b3869de-5b90-44f6-8a6d-c4e078e3b666)

However, in the vast majority of spaced repetition software used individually, given the same review history, the algorithm presents minor variations in intervals, coupled with a scarcity of data, rendering the estimation of the retention rate imprecise. Incorporating parameters on the forgetting curve might lead the algorithm to learn the characteristics of a specific retention rate, which would perform inadequately in extrapolation.

**Comment by Expertium** [comment]
I'll recommend closing this issue since the new version is getting released, and this issue hasn't been used for a long time.
